### ISSUE TITLE
Implement usage of composer autoloading

### ISSUE THREAD

**Issue by pfalomo** [ClassicPress/ClassicPress#247]
As business developer i use often the composer autoloading feature, especially when adopting a private repository like satis. This is a missing feature on wordpress since the fact it has been started in "old php times" when composer was just a dream.

I think it would be a brilliant feature enabling the composer autoloading including hooks from [wpackagist](https://wpackagist.org/).

NOTE: bedrock already does it https://roots.io/using-composer-with-wordpress/ in some sort of way

**Comment by Mte90** [comment]
Hi we already implemented that :-)
Check at https://github.com/ClassicPress/ClassicPress/issues/218 we have also docs and as you can see in the repo there is already the composer.json file

**Comment by pfalomo** [comment]
@Mte90 it seems it only supports wpackagist. I meant to autoload even other php packages (eg an xml parser etc..). Is it included?

**Comment by Mte90** [comment]
Yes because composer as default has already packagist as repo, in that composer file we added also wpackagist to speed up for who use the plugin in that way.

**Comment by pfalomo** [comment]
alright, i'm trying to get the last build locally. Looking further for this. Thank you!

**Comment by nylen** [comment]
This issue is for enabling composer **autoloading**, not just composer.  @pfalomo if you are able to help outline what we need to do here, or if you can help with an example implementation, that would be much appreciated.

**Comment by pierre-dargham** [comment]
Hi,

After installing ClassicPress via Composer, enabling composer autoloading of third party packages requires no specific implementation.

Just as in any composer project, you just need to load composer file `vendor/autoload.php` in your application bootstrap.

Example : https://github.com/pierre-dargham/classicpress-composer-boilerplate/blob/master/wp-config.php#L99

**Comment by pfalomo** [comment]
@pierre-dargham , you are right. But i was trying to install the latest version of classicpress days ago and i wasn't able to clone the beta. Just alpha (where it has not this implementation). 
I'm not able to test it now again. I'll try soon.

**Comment by pfalomo** [comment]
@nylen what @pierre-dargham suggests it's a good way to implement (similar to [bedrock](https://roots.io/bedrock/))

**Comment by pfalomo** [comment]
I'd suggest to include a conditional `require_once`

```php
//maybe end of wp-config.php before calling wp-settings.php
if(file_exists( __DIR__ . '/vendor/autoload.php'){
   require_once __DIR__ . '/vendor/autoload.php';
}
```

because if you install CP manually after download from site you have no vendor but you have composer.json and you may want to start using packages.

Note: `wp-config.php` might not be the best point to start autoloading.

**Comment by pierre-dargham** [comment]
@pfalomo thank you for the feedbacks ! A few answers :

> I'd suggest to include a conditional require_once

- Installing CP using composer, you will _necessarily_ end up with a `vendor/autoload.php` file. Despite it would be technically possible to install CP manually from a zip archive and manage the core without composer, while having composer dependencies, it would not make much sense, according to me. Either you don't use composer at all, either you properly manage all depdendencies (according the core) through it.

> wp-config.php might not be the best point to start autoloading

- Agreed, but we have to stay consistent with WordPress bootstrap, and this boilerplate wants to show the minimal steps, so I don't want to make a new bootstrap / configuration structure here. I'll leave that to more complete stacks

By the way, note that it's just a personal repository I started as an example for now, not an official CP repo. If you have specific comments about it, it may be better to do them on https://github.com/pierre-dargham/classicpress-composer-boilerplate ;)

**Comment by pfalomo** [comment]
![image](https://user-images.githubusercontent.com/44964509/49388771-08e6e380-f725-11e8-90c2-135933a94a38.png)

This is mainly insipired to Laravel Framework (as someone may have noticed).
the `cp/` folder should be autoloaded. It is just a dream? Can you see it?

**Comment by pfalomo** [comment]
@pierre-dargham

>By the way, note that it's just a personal repository I started as an example for now, not an official CP repo. If you have specific comments about it, it may be better to do them on https://github.com/pierre-dargham/classicpress-composer-boilerplate ;)

Yep, noticed after i posted. By the way i totally share your point of you and proposed a [petition](https://petitions.classicpress.net/posts/117/overrides-improvements-oop-scalability) time ago

**Comment by pattonwebz** [comment]
@szepeviktor this may be something you are interested in looking at as I know you have already done a lot of investigations about how to autoload core.

**Comment by szepeviktor** [comment]
Thank you for mentioning me.

Here is how we can approach this transition.

1. Dump brute force - [a shell implementation](https://github.com/szepeviktor/wordpress-autoloaded/blob/master/bin/remove-class-loading.sh)
1. Incrementally by separating side-effects (see PSR-1) leaving only clean class files and files containing only functions

We may consider class files name format, using namespaces. See [Themosis](https://github.com/themosis/framework/tree/2.0/src) on making WP OOP

**Comment by ginsterbusch** [comment]
> We may consider class file name format and using namespaces. See [Themosis](https://github.com/themosis/framework/tree/2.0/src) on making WP OOP

This is rather OT. but:
We have been in talks about this for other things, eg. my proposal of an improved media library. My idea was to use a namespace for the experimental plugin, to avoid possible clashing with other plugins and classes (clashing as in `FATAL ERROR` :D), and to easier separate several features to be.

cu, w0lf.

**Comment by Mte90** [comment]
I am worried on the case 2 for the backward compatibility because the core is very spaghetti code.
Maybe we can do it for the wp-includes and not for the admin stuff.

**Comment by szepeviktor** [comment]
Any progress is progress.

**Comment by bahiirwa** [comment]
> I am worried on the case 2 for the backward compatibility because the core is very spaghetti code.

Do we have to be backward compatible for the next version of CP? I thought version 1 was the LTS. We can make things modern for the next version after the directory.

**Comment by Mte90** [comment]
For version 2 yes, the point is that with this kind of changes we can break the compatibility with a lot of plugins.
If we break compatibilities this can create a lot of issues on adopting CP.
Also it is another point that we never discussed about what means for CP break compatibility in version 1 and 2 also for backporting changes.

**Comment by mattyrob** [comment]
This issue seems quite stale and has not gained any traction or any proposed PR.

I suggest we either close this and it can be re-opened if needed or we add the proposed snippet of code as above into either `wp-config.php` or `wp-load.php`.

**Comment by KTS915** [comment]
I'm comfortable with closing this. It's already possible for anyone who wants to do so to install ClassicPress via composer using tools provided by either John P. Bloch or Bedrock, and it's not hard to find tutorials on how to make the process work. But they both require that CP is installed in a subdirectory, which will cause issues on some hosts and break some plugins.

If someone would like the process built-in, then they should submit a PR.

**Comment by mattyrob** [comment]
Almost a year since last comments, this is now stale. Closing.
